app/models/workers_scaling/experiment_resources_interface.rb

```diff
@@ -0,0 +1,221 @@
+        return starting_workers + initializing_workers if amount <= 0
+
+        params.merge! infrastructure[:params]
+        # Time limit is 1 year in minutes
+        params[:time_limit] = 60 if params[:time_limit].nil?
+
+        params = ActiveSupport::HashWithIndifferentAccess.new(params)
+        get_facade_for(infrastructure[:name])
+          .start_simulation_managers(@user_id, amount, @experiment.id.to_s, params)
+          .map(&:sm_uuid)
+          .concat(starting_workers + initializing_workers)
+      rescue InvalidCredentialsError, NoCredentialsError
+        # TODO inform user about credentials error
+        raise
+      end
+    end
+
+    ##
+    # Marks worker with given sm_uuid to be deleted after finishing given number of simulations
+    # If overwrite is set to false, worker will not be marked if simulations_left field is already set,
+    # otherwise new number will be set in all cases
+    # Using overwrite=true may result in unintentional resetting simulations_left field,
+    # effectively prolonging lifetime of worker, therefore default value is false
+    def limit_worker_simulations(sm_uuid, simulations_left, overwrite=false)
+      worker = get_worker_record_by_sm_uuid(sm_uuid)
+      unless worker.nil?
+        worker.simulations_left = simulations_left if worker.simulations_left.blank? or overwrite
+        worker.save
+      end
+    end
+
+    ##
+    # Stops worker after completing its current simulation execution
+    def soft_stop_worker(sm_uuid)
+      limit_worker_simulations(sm_uuid, 1)
+    end
+
+    ##
+    # Returns list of workers records for infrastructure
+    def get_workers_records_list(infrastructure, params = {})
+      get_workers_records_cursor(infrastructure, params).to_a
+    end
+
+    ##
+    # Returns workers records count for infrastructure
+    def get_workers_records_count(infrastructure, params = {})
+      get_workers_records_cursor(infrastructure, params).count
+    end
+
+    ##
+    # Returns overall Workers count for Experiment matching given params
+    def count_all_workers(params = {})
+      get_available_infrastructures
+          .flat_map { |infrastructure| get_workers_records_count(infrastructure, params) }
+          .reduce(0) { |sum, count| sum + count }
+    end
+
+    ##
+    # Returns worker record for given sm_uuid
+    def get_worker_record_by_sm_uuid(sm_uuid)
+      get_available_infrastructures.flat_map do |infrastructure|
+        get_workers_records_cursor(infrastructure, cond: {sm_uuid: sm_uuid}).first
+      end .first
+    end
+
+    ##
+    # Yields workers for given records
+    # Usage:
+    #   yield_workers(records) do |worker|
+    #      worker.<some method>
+    #   end
+    # Workers commands: restart, stop, destroy_record
+    def yield_workers(records)
+      records.each do |record|
+        get_facade_for(record.infrastructure).yield_simulation_manager(record) {|worker| yield worker}
```

Comment:
W tym miejscu mam nietypową propozycję rozwiązania - usunąć tę metodę. Nie używamy jej, wiele razy przekonaliśmy się, że nie powinniśmy - natychmiastowy stop jest złym pomysłem, z punktu widzenia algorytmu nie ma sensu restart ani zniszczenie rekordu.

Jeżeli chcemy jednak jej używać, to będą to pewnie jakieś skrajne przypadki, dla których nie warto bawić się optymalizacją. Nie jest to proste, bo yield_simulation_managers przyjmuje bardzo niemiłe *args, które okazują się być takim zestawem: user_id, experiment_id, params. Nie ma tutaj miejsca na podanie konkretnej listy rekordów, z których chcieliśmy stworzyć workera i wywołać na nim operację.

tl;dr: Proponuję to usunąć albo zignorować.